### ISSUE TITLE
Crash on right clicking in Solution window

### ISSUE THREAD

**Issue by mrward** [mono/monodevelop#7046]
Release ID: 8.0.0.1344
Git revision: b7efc0cb39fe50ffde40506da2df3d62f11a8528
Build date: 2019-02-03 16:55:27+00
Build branch: master
Xamarin extensions: 956d4c6701a2c8224a2efe99de3cb91176423155

Visual Studio Enterprise 2019 Preview for Mac
Version 8.0 Preview (8.0 build 1344)
Xamarin.Mac 5.6.0.1 (d16-0 / 2e9867a4)
Mono 5.18.0.248 (2018-08/a4956c837e1) (64-bit)

Repro:

1. Create C# library project.
2. Right click the project and select Add - New Folder.
3. Click the cancel button on the dialog.
4. Right click the project again.
5. IDE crashes.

See a similar crash if you trigger the NuGet accept licenses dialog, say by installing Mvc, then after you close that dialog you try to right click the project in the Solution window.

Is this some sort of XWT - Xamarin.Mac problem? Seems like using native toolkit with XWT is leaving a handle/dialog around. Did not see this crash testing the New Folder dialog, but I have Xamarin.Mac 5.6.0.2 installed so would have been using that with VS Mac in a local build.

```
An unhandled exception has occurred. Terminating Visual Studio? False
ObjCRuntime.RuntimeException: Failed to marshal the Objective-C object 0x7fcc598754f0 (type: MonoDevelop_MacIntegration_ThemedMacDialogBackend). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance (because the type 'MonoDevelop.MacIntegration.ThemedMacDialogBackend' does not have a constructor that takes one IntPtr argument).
  at ObjCRuntime.Runtime.MissingCtor (System.IntPtr ptr, System.IntPtr klass, System.Type type, ObjCRuntime.Runtime+MissingCtorResolution resolution, System.IntPtr selector, System.IntPtr method) [0x000e2] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/ObjCRuntime/Runtime.cs:1113 
  at ObjCRuntime.Runtime.ConstructNSObject[T] (System.IntPtr ptr, System.Type type, ObjCRuntime.Runtime+MissingCtorResolution missingCtorResolution, System.IntPtr selector, System.IntPtr method) [0x0002a] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/ObjCRuntime/Runtime.cs:1142 
  at ObjCRuntime.Runtime.GetINativeObject[T] (System.IntPtr ptr, System.Boolean owns) [0x00139] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/ObjCRuntime/Runtime.cs:1469 
  at Foundation.NSArray.UnsafeGetItem[T] (System.IntPtr handle, System.nuint index) [0x00029] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/Foundation/NSArray.cs:345 
  at Foundation.NSArray.ArrayFromHandle[T] (System.IntPtr handle) [0x0002e] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/Foundation/NSArray.cs:251 
  at AppKit.NSApplication.get_Windows () [0x00012] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/AppKit/NSApplication.g.cs:6802 
  at MonoDevelop.MacInterop.GtkQuartz+<GetToplevels>d__3.MoveNext () [0x00017] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MacPlatform/MacInterop/GtkQuartz.cs:61 
  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].MoveNext () [0x0004e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Where.cs:140 
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00039] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/AnyAll.cs:36 
  at MonoDevelop.MacIntegration.MacPlatformService.IsModalDialogRunning () [0x00014] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/addins/MacPlatform/MacPlatform.cs:1088 
  at MonoDevelop.Ide.DesktopService.IsModalDialogRunning () [0x00000] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/DesktopService.cs:387 
  at MonoDevelop.Components.Mac.MDMenuItem.get_IsGloballyDisabled () [0x00000] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:100 
  at MonoDevelop.Components.Mac.MDMenuItem.SetItemValues (AppKit.NSMenuItem item, MonoDevelop.Components.Commands.CommandInfo info, System.Boolean disabledVisible, System.String overrideLabel) [0x0000e] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:210 
  at MonoDevelop.Components.Mac.MDMenuItem.Update (AppKit.NSMenu parent, System.Int32& index, MonoDevelop.Components.Commands.CommandInfo info) [0x00008] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:141 
  at MonoDevelop.Components.Mac.MDMenuItem.Update (MonoDevelop.Components.Mac.MDMenu parent, System.Int32& index) [0x00080] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenuItem.cs:125 
  at MonoDevelop.Components.Mac.MDMenu.UpdateCommands () [0x00037] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenu.cs:157 
  at MonoDevelop.Components.Mac.MDSubMenuItem.Update (MonoDevelop.Components.Mac.MDMenu parent, System.Int32& index) [0x00000] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDSubMenuItem.cs:48 
  at MonoDevelop.Components.Mac.MDMenu.UpdateCommands () [0x00037] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenu.cs:157 
  at MonoDevelop.Components.Mac.MDMenu.MenuNeedsUpdate (AppKit.NSMenu menu) [0x00019] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Mac/MDMenu.cs:199 
  at (wrapper managed-to-native) ObjCRuntime.Messaging.void_objc_msgSend_IntPtr_IntPtr_IntPtr(intptr,intptr,intptr,intptr,intptr)
  at AppKit.NSMenu.PopUpContextMenu (AppKit.NSMenu menu, AppKit.NSEvent theEvent, AppKit.NSView view) [0x0003a] in /Library/Frameworks/Xamarin.Mac.framework/Versions/5.6.0.1/src/Xamarin.Mac/AppKit/NSMenu.g.cs:1807 
  at MonoDevelop.Components.ContextMenuExtensionsMac+<>c__DisplayClass4_0.<ShowContextMenu>b__0 (System.Object o, System.EventArgs args) [0x0013c] in /Users/vsts/agent/2.146.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenuExtensionsMac.cs:107 
  at Gtk.Application+InvokeProxy.HandlerInternal (System.IntPtr data) [0x0001d] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:199 

```

> VS bug [#785490](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/785490)

**Comment by mrward** [comment]
Crashes with release-8.0 build too. The information below says master but it is a release-8.0 build. It is using Xamarin.Mac 5.6.0.2 which master is not.

```
Visual Studio Enterprise 2019 Preview for Mac
Version 8.0 Preview (8.0 build 1805)
Installation UUID: 19384e93-709f-410c-9c15-434da9fba83e
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.6.0.2 (d16-0 / 040682909)

	Package version: 518000248

Mono Framework MDK
Runtime:
	Mono 5.18.0.248 (2018-08/a4956c837e1) (64-bit)
	Package version: 518000248


Xamarin.Mac
Version: 5.6.0.2 (Visual Studio Enterprise)
Hash: 040682909
Branch: d16-0
Build date: 2018-12-14 06:23:20-0500

Build Information
Release ID: 800001805
Git revision: 976f9866afea4b6a99d4e875e9bf9b060a9f4024
Build date: 2019-01-31 18:09:53+00
Build branch: master
Xamarin extensions: 16ded92f644e8fcbad9a798adaab198a198dc651

```

**Comment by mrward** [comment]
VS Mac 7.8.0.1555 (which is a bit old) seems OK.

**Comment by sevoku** [comment]
This is again the Xamarin.Mac issue with resurrecting NSWindows (NSPanel doesn't have that issue, so the file dialogs and also native tooltips are fine). This is also unrelated to Xwt and can happen with any NSWindow. This is basically what happens:
- NSWindow gets disposed and collected
- the unmanaged NSWindow instance is kept alive for a couple of cycles (by AppKit)
- if [NSApplication.SharedApplication.Windows](https://github.com/mono/monodevelop/blob/ffb3143905a3db94f2384f56f000cbfd7e08b247/main/src/addins/MacPlatform/MacInterop/GtkQuartz.cs#L61) is accessed while the native window is alive:
  - Xamarin.Mac tries to create the managed object for the dead window and fails

There are two ways to solve this:

**Hack**
Add the IntPtr constructor to the Xwt backend and allow resurrection. This could potentially leak, since we would resurrect windows after they should have been released. I'm not sure about possible issues with that state of unmanaged NSwindows. But it would at least solve the crasher.

**Right fix**
Track all native windows manually in MD and never access `NSApplication.SharedApplication.Windows` (we could create an analyzer for that). The problem with this approach is how to reliably track ALL native window creation/destruction (even from extensions!). We could use `NSWindowWillCloseNotification` to track closed windows, but I don't know what we could use to track creation. Maybe we could disallow using NSWindow in extensions somehow and provide an API?

@Therzok we discussed this on Slack a month ago, do you have an idea?

**Comment by Therzok** [comment]
I agree with the right fix. We can insert such an analyzer via AddinMaker.

Similarly, since we ship Xamarin.Mac, we can completely remove the method from existence via IL processing.

**Comment by netonjm** [comment]
+1 to right fix

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.0.0.1456 (master)

Author: Vsevolod Kukol
Commit: 294baa6e723eccc0ff0890d93198b62f73081333 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-17] Comment by **xamarinc**



Fixed in version 8.0.0.1456 (master)



Pull Request #7090 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 437dc38210a98901afbeb51756119594ce67ba60 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-18] Comment by **xamarinc**



Fixed in version 8.0.0.2176 (release-8.0)



Pull Request #7090 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 437dc38210a98901afbeb51756119594ce67ba60 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-29] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #7090 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 437dc38210a98901afbeb51756119594ce67ba60 (xamarin/monodevelop)
